File: runtimes/r/R/dictionary.R
decode_trained_dictionary_payload <- function(payload) {
  suppressPackageStartupMessages(requireNamespace("bit64", quietly = TRUE))
  reader <- new_reader(payload)
  n <- reader$read_count()
  values <- vector("character", n)
  for (i in seq_len(n)) values[[i]] <- reader$read_string()
  if (!reader$is_eof()) twilic_stop(invalid_data("trained dictionary payload trailing bytes"))
  values
}

encode_trained_dictionary_block <- function(values, dictionary) {
  if (!length(values)) {
    out <- raw_append_byte(new_buffer(), 0L)
    return(list(encode_varuint(0L, out), TRUE, NULL))
  }
  by_value <- setNames(seq_along(dictionary) - 1L, dictionary)
  ids <- numeric()
  for (value in values) {
    ref_id <- by_value[[value]]
    if (is.null(ref_id)) return(list(NULL, FALSE, NULL))
    ids <- c(ids, ref_id)
  }
  raw <- raw_append_byte(new_buffer(), 0L)
  raw <- encode_varuint(length(ids), raw)
  for (ref_id in ids) raw <- encode_varuint(ref_id, raw)
  max_id <- if (length(ids)) max(ids) else 0
  bit_width <- if (max_id == 0) 0L else bit_length64(max_id)
  packed <- pack_fixed_width_u64(ids, bit_width, new_buffer())
  bitpacked <- raw_append_byte(new_buffer(), 1L)
  bitpacked <- encode_varuint(length(ids), bitpacked)
  bitpacked <- raw_append_byte(bitpacked, bit_width)
  bitpacked <- raw_append_bytes(bitpacked, packed)
  if (length(bitpacked) < length(raw)) return(list(bitpacked, TRUE, NULL))
  list(raw, TRUE, NULL)
}

decode_trained_dictionary_block <- function(block, dictionary) {
  reader <- new_reader(block)
  mode <- reader$read_u8()
  n <- reader$read_count()
  if (mode == 0L) {
    ids <- vapply(seq_len(n), function(i) reader$read_varuint(), numeric(1))
  } else if (mode == 1L) {
    bit_width <- reader$read_u8()
    remaining <- length(block) - reader$position()
    packed <- reader$read_exact(remaining)
    ids <- unpack_fixed_width_u64(packed, n, bit_width)
  } else {
    twilic_stop(invalid_data("trained dictionary block mode"))
  }
  if (!reader$is_eof()) twilic_stop(invalid_data("trained dictionary block trailing bytes"))
  out <- character()
  for (ref_id in ids) {
    if (ref_id >= length(dictionary)) twilic_stop(invalid_data("trained dictionary block id"))
    out <- c(out, dictionary[[ref_id + 1L]])
  }
  out
}

wide_u128_mask <- function(width) {
  if (width == 0L) return(list(lo = 0L, hi = 0L))
  if (width >= 64L) return(list(lo = -1L, hi = -1L))
  if (width == 32L) return(list(lo = -1L, hi = 0L))
  if (width < 32L) return(list(lo = bitwShiftL(1L, width) - 1L, hi = 0L))
  list(lo = -1L, hi = bitwShiftL(1L, width - 32L) - 1L)
}

wide_u128_and <- function(a, b) list(lo = bitwAnd(a$lo, b$lo), hi = bitwAnd(a$hi, b$hi))
wide_u128_or <- function(a, b) list(lo = bitwOr(a$lo, b$lo), hi = bitwOr(a$hi, b$hi))
wide_u128_shr <- function(a, n) {
  if (n == 0L) return(a)
  if (n >= 128L) return(list(lo = 0, hi = 0))
  if (n < 64L) {
    return(list(
      lo = bitwAnd(bitwOr(bitwShiftR(a$lo, n), bitwShiftL(a$hi, 64L - n)), bitwShiftL(1L, 64) - 1),
      hi = bitwShiftR(a$hi, n)
    ))
  }
  list(lo = bitwShiftR(a$hi, n - 64L), hi = 0)
}

wide_u128_shl <- function(a, n) {
  if (n == 0L) return(a)
  if (n >= 128L) return(list(lo = 0, hi = 0))
  if (n < 64L) {
    return(list(
      lo = bitwAnd(bitwShiftL(a$lo, n), bitwShiftL(1L, 64) - 1),
      hi = bitwAnd(bitwOr(bitwShiftL(a$hi, n), bitwShiftR(a$lo, 64L - n)), bitwShiftL(1L, 64) - 1)
    ))
  }
  list(lo = 0, hi = bitwAnd(bitwShiftL(a$lo, n - 64L), bitwShiftL(1L, 64) - 1))
}

pack_fixed_width_u64 <- function(values, width, out) {
  if (width > 64L) twilic_stop(invalid_data("fixed-width u64 bit width"))
  if (width == 0L) {
    if (any(values != 0)) twilic_stop(invalid_data("fixed-width u64 value overflow"))
    return(out)
  }
  acc <- list(lo = 0, hi = 0)
  acc_bits <- 0L
  for (value in values) {
    if (width < 64L && width > 0L && value >= wide_u128_mask(width)$lo) {
      twilic_stop(invalid_data("fixed-width u64 value overflow"))
    }
    acc <- wide_u128_or(acc, wide_u128_shl(list(lo = as.numeric(value), hi = 0), acc_bits))
    acc_bits <- acc_bits + width
    while (acc_bits >= 8L) {
      out <- raw_append_byte(out, bitwAnd(acc$lo, 0xFFL))
      acc <- wide_u128_shr(acc, 8L)
      acc_bits <- acc_bits - 8L
    }
  }
  if (acc_bits > 0L) out <- raw_append_byte(out, bitwAnd(acc$lo, 0xFFL))
  out
}

unpack_fixed_width_u64 <- function(data, count, width) {
  if (width > 64L) twilic_stop(invalid_data("fixed-width u64 bit width"))
  if (width == 0L) {
    if (any(as.integer(data) != 0L)) twilic_stop(invalid_data("fixed-width u64 trailing bytes"))
    return(rep(0, count))
  }
  out <- numeric(count)
  acc <- list(lo = 0L, hi = 0L)
  acc_bits <- 0L
  idx <- 1L
  mask <- wide_u128_mask(width)
  for (i in seq_len(count)) {
    while (acc_bits < width) {
      if (idx > length(data)) twilic_stop(invalid_data("fixed-width u64 underflow"))
      acc <- wide_u128_or(acc, wide_u128_shl(list(lo = as.integer(data[idx]), hi = 0L), acc_bits))
      idx <- idx + 1L
      acc_bits <- acc_bits + 8L
    }
    out[i] <- wide_u128_and(acc, mask)$lo
    acc <- wide_u128_shr(acc, width)
    acc_bits <- acc_bits - width
  }
  if (acc$lo != 0L || acc$hi != 0L) twilic_stop(invalid_data("fixed-width u64 trailing bytes"))
  for (j in idx:length(data)) {
    if (as.integer(data[j]) != 0L) twilic_stop(invalid_data("fixed-width u64 trailing bytes"))
  }
  out
}

apply_dictionary_references <- function(state, columns) {
  for (i in seq_along(columns)) {
    column <- columns[[i]]
    if (column$values$kind != ElementTypeSTRING) next
    values <- unlist(column$values$strings)
    if (length(values) < 16L) next
    unique_vals <- unique(values)
    if (length(unique_vals) / length(values) > 0.5) next
    if (!(column$codec %in% c(VectorCodecDICTIONARY, VectorCodecSTRING_REF))) next
    dict_id <- allocate_dictionary_id(state)
    payload <- new_buffer()
    keys <- sort(unique_vals)
    payload <- encode_varuint(length(keys), payload)
    for (item in keys) payload <- encode_string(item, payload)
    profile <- list(
      version = 1L,
      hash = dictionary_payload_hash(payload),
      expires_at = 0L,
      fallback = DictionaryFallbackFailFast
    )
    if (state$options$unknown_reference_policy == UnknownReferencePolicyStatelessRetry) {
      profile$fallback <- DictionaryFallbackStatelessRetry
    }
    state$dictionaries[[as.character(dict_id)]] <- payload
    state$dictionary_profiles[[as.character(dict_id)]] <- profile
    column$dictionary_id <- dict_id
    columns[[i]] <- column
  }
  columns
}

mul32 <- function(x, y) {
  x <- as.numeric(x) %% 4294967296
  y <- as.numeric(y) %% 4294967296
  x0 <- x %% 65536
  x1 <- floor(x / 65536)
  y0 <- y %% 65536
  y1 <- floor(y / 65536)
  z00 <- x0 * y0
  z01 <- x0 * y1
  z10 <- x1 * y0
  z11 <- x1 * y1
  mid <- floor(z00 / 65536) + z01 + z10
  lo <- (z00 %% 65536) + (mid %% 65536) * 65536
  hi <- z11 + floor(mid / 65536) + floor(lo / 4294967296)
  list(lo = lo %% 4294967296, hi = hi %% 4294967296)
}

add64 <- function(lo, hi, dlo = 0, dhi = 0) {
  lo <- lo + dlo
  hi <- hi + dhi + floor(lo / 4294967296)
  list(lo = lo %% 4294967296, hi = hi %% 4294967296)
}

u64_mul_mod_prime <- function(a_lo, a_hi, m) {
  m_lo <- m %% 4294967296
  m_hi <- floor(m / 4294967296)
  t0 <- mul32(a_lo, m_lo)
  t1 <- mul32(a_lo, m_hi)
  t2 <- mul32(a_hi, m_lo)
  r <- add64(t0$lo, t0$hi)
  r <- add64(r$lo, r$hi, dhi = t1$lo)
  add64(r$lo, r$hi, dhi = t2$lo)
}

FNV1A_OFFSET_RAW <- as.raw(c(0x25, 0x34, 0x22, 0x84, 0xe4, 0x9c, 0xf2, 0xcb))

u32_from_raw4 <- function(r) {
  v <- bit64::as.integer64(as.integer(r[[1L]])) +
    bit64::as.integer64(as.integer(r[[2L]])) * bit64::as.integer64(256) +
    bit64::as.integer64(as.integer(r[[3L]])) * bit64::as.integer64(65536) +
    bit64::as.integer64(as.integer(r[[4L]])) * bit64::as.integer64(16777216)
  as.numeric(as.character(v))
}

raw8_to_limbs <- function(raw8) {
  list(lo = u32_from_raw4(raw8[1:4]), hi = u32_from_raw4(raw8[5:8]))
}

limbs_to_raw8 <- function(lo, hi) {
  c(u32_to_le_raw(lo), u32_to_le_raw(hi))
}

fnv1a64 <- function(payload) {
  payload <- as_raw_input(payload)
  h <- FNV1A_OFFSET_RAW
  prime <- 1099511628211
  for (b in as.integer(payload)) {
    h[[1L]] <- as.raw(bitwXor(as.integer(h[[1L]]), b))
    limbs <- raw8_to_limbs(h)
    nxt <- u64_mul_mod_prime(limbs$lo, limbs$hi, prime)
    h <- limbs_to_raw8(nxt$lo, nxt$hi)
  }
  u64_from_le_bytes(h)
}

dictionary_payload_hash <- function(payload) fnv1a64(payload)

File: runtimes/swift/Sources/Twilic/Core/Dictionary.swift
import Foundation

public struct EncodedDictionaryBlock: Sendable {
    public var block: Data?
    public var ok: Bool
}

enum DictionaryCore {
    static func decodeTrainedDictionaryPayload(_ payload: Data) throws -> [String] {
        let reader = Wire.newReader(payload)
        let n = try reader.readVaruint()
        var values: [String] = []
        for _ in 0 ..< Int(n) {
            values.append(try reader.readString())
        }
        if !reader.isEOF {
            throw TwilicErrors.invalidData("trained dictionary payload trailing bytes")
        }
        return values
    }

    static func encodeTrainedDictionaryBlock(
        _ values: [String],
        _ dictionary: [String]
    ) throws -> EncodedDictionaryBlock {
        if values.isEmpty {
            var out = Data()
            out.append(0)
            Wire.encodeVaruint(0, &out)
            return EncodedDictionaryBlock(block: out, ok: true)
        }
        var byValue: [String: UInt64] = [:]
        for (idx, value) in dictionary.enumerated() {
            byValue[value] = UInt64(idx)
        }
        var ids: [UInt64] = []
        for value in values {
            guard let refID = byValue[value] else {
                return EncodedDictionaryBlock(block: nil, ok: false)
            }
            ids.append(refID)
        }
        var raw = Data()
        raw.append(0)
        Wire.encodeVaruint(UInt64(ids.count), &raw)
        for refID in ids {
            Wire.encodeVaruint(refID, &raw)
        }
        let maxID = ids.max() ?? 0
        let bitWidth = maxID == 0 ? 0 : (64 - maxID.leadingZeroBitCount)
        var packed = Data()
        try packFixedWidthU64(ids, bitWidth, &packed) // swiftlint:disable
        var bitpacked = Data()
        bitpacked.append(1)
        Wire.encodeVaruint(UInt64(ids.count), &bitpacked)
        bitpacked.append(UInt8(bitWidth))
        bitpacked.append(packed)
        if bitpacked.count < raw.count {
            return EncodedDictionaryBlock(block: bitpacked, ok: true)
        }
        return EncodedDictionaryBlock(block: raw, ok: true)
    }

    static func decodeTrainedDictionaryBlock(_ block: Data, _ dictionary: [String]) throws -> [String] {
        let reader = Wire.newReader(block)
        let mode = try reader.readU8()
        let n = try reader.readCount()
        let ids: [UInt64]
        if mode == 0 {
            ids = try (0 ..< n).map { _ in try reader.readVaruint() }
        } else if mode == 1 {
            let bitWidth = Int(try reader.readU8())
            let remaining = block.count - reader.position
            let packed = try reader.readExact(remaining)
            ids = try unpackFixedWidthU64(packed, n, bitWidth)
        } else {
            throw TwilicErrors.invalidData("trained dictionary block mode")
        }
        if !reader.isEOF {
            throw TwilicErrors.invalidData("trained dictionary block trailing bytes")
        }
        var out: [String] = []
        for refID in ids {
            guard refID < UInt64(dictionary.count) else {
                throw TwilicErrors.invalidData("trained dictionary block id")
            }
            out.append(dictionary[Int(refID)])
        }
        return out
    }

    private struct WideU128 {
        var lo: UInt64
        var hi: UInt64

        static func fromU64(_ v: UInt64) -> WideU128 { WideU128(lo: v, hi: 0) }

        static func mask(_ width: Int) -> WideU128 {
            if width == 64 { return WideU128(lo: UInt64.max, hi: UInt64.max) }
            if width == 0 { return WideU128(lo: 0, hi: 0) }
            if width <= 64 { return WideU128(lo: (1 << width) &- 1, hi: 0) }
            return WideU128(lo: UInt64.max, hi: (1 << (width - 64)) &- 1)
        }

        func isZero() -> Bool { lo == 0 && hi == 0 }

        func and(_ other: WideU128) -> WideU128 {
            WideU128(lo: lo & other.lo, hi: hi & other.hi)
        }

        func or(_ other: WideU128) -> WideU128 {
            WideU128(lo: lo | other.lo, hi: hi | other.hi)
        }

        func shl(_ n: Int) -> WideU128 {
            if n == 0 { return self }
            if n >= 128 { return WideU128(lo: 0, hi: 0) }
            if n < 64 {
                let newHi = (hi << n) | (lo >> (64 - n))
                let newLo = lo << n
                return WideU128(lo: newLo, hi: newHi)
            }
            return WideU128(lo: 0, hi: lo << (n - 64))
        }

        func shr(_ n: Int) -> WideU128 {
            if n == 0 { return self }
            if n >= 128 { return WideU128(lo: 0, hi: 0) }
            if n < 64 {
                let newLo = (lo >> n) | (hi << (64 - n))
                let newHi = hi >> n
                return WideU128(lo: newLo, hi: newHi)
            }
            return WideU128(lo: hi >> (n - 64), hi: 0)
        }
    }

    private static func packFixedWidthU64(_ values: [UInt64], _ width: Int, _ out: inout Data) throws {
        if width > 64 { throw TwilicErrors.invalidData("fixed-width u64 bit width") }
        if width == 0 {
            for value in values where value != 0 {
                throw TwilicErrors.invalidData("fixed-width u64 value overflow")
            }
            return
        }
        var acc = WideU128(lo: 0, hi: 0)
        var accBits = 0
        for value in values {
            if width < 64 && value >> width != 0 {
                throw TwilicErrors.invalidData("fixed-width u64 value overflow")
            }
            acc = acc.or(WideU128.fromU64(value).shl(accBits))
            accBits += width
            while accBits >= 8 {
                out.append(UInt8(acc.lo & 0xFF))
                acc = acc.shr(8)
                accBits -= 8
            }
        }
        if accBits > 0 {
            out.append(UInt8(acc.lo & 0xFF))
        }
    }

    private static func unpackFixedWidthU64(_ data: Data, _ count: Int, _ width: Int) throws -> [UInt64] {
        if width > 64 { throw TwilicErrors.invalidData("fixed-width u64 bit width") }
        if width == 0 {
            for b in data where b != 0 {
                throw TwilicErrors.invalidData("fixed-width u64 trailing bytes")
            }
            return Array(repeating: 0, count: count)
        }
        var out: [UInt64] = []
        var acc = WideU128(lo: 0, hi: 0)
        var accBits = 0
        var idx = 0
        let mask = WideU128.mask(width)
        for _ in 0 ..< count {
            while accBits < width {
                guard idx < data.count else {
                    throw TwilicErrors.invalidData("fixed-width u64 underflow")
                }
                acc = acc.or(WideU128.fromU64(UInt64(data[idx])).shl(accBits))
                idx += 1
                accBits += 8
            }
            out.append(acc.and(mask).lo)
            acc = acc.shr(width)
            accBits -= width
        }
        if !acc.isZero() {
            throw TwilicErrors.invalidData("fixed-width u64 trailing bytes")
        }
        for j in idx ..< data.count where data[j] != 0 {
            throw TwilicErrors.invalidData("fixed-width u64 trailing bytes")
        }
        return out
    }

    static func applyDictionaryReferences(_ state: SessionState, _ columns: [Column]) {
        for column in columns {
            guard column.values.kind == .string else { continue }
            let values = column.values.strings
            if values.count < 16 { continue }
            let unique = Set(values)
            if Double(unique.count) / Double(values.count) > 0.5 { continue }
            guard column.codec == .dictionary || column.codec == .stringRef else { continue }
            let dictID = allocateDictionaryID(state)
            var payload = Data()
            let keys = unique.sorted()
            Wire.encodeVaruint(UInt64(keys.count), &payload)
            for item in keys {
                Wire.encodeString(item, &payload)
            }
            var profile = DictionaryProfile(
                version: 1,
                hash: dictionaryPayloadHash(payload),
                expiresAt: 0,
                fallback: .failFast
            )
            if state.options.unknownReferencePolicy == .statelessRetry {
                profile.fallback = .statelessRetry
            }
            state.dictionaries[dictID] = payload
            state.dictionaryProfiles[dictID] = profile
            var col = column
            col.dictionaryID = dictID
        }
    }

    static func dictionaryPayloadHash(_ payload: Data) -> UInt64 {
        fnv1a64(payload)
    }

    static func fnv1a64(_ payload: Data) -> UInt64 {
        var h: UInt64 = 0xCBF29CE484222325
        for b in payload {
            h ^= UInt64(b)
            h = h &* 0x100000001B3
        }
        return h
    }
}

public func decodeTrainedDictionaryPayload(_ payload: Data) throws -> [String] {
    try DictionaryCore.decodeTrainedDictionaryPayload(payload)
}

public func encodeTrainedDictionaryBlock(
    _ values: [String],
    _ dictionary: [String]
) throws -> EncodedDictionaryBlock {
    try DictionaryCore.encodeTrainedDictionaryBlock(values, dictionary)
}

public func decodeTrainedDictionaryBlock(_ block: Data, _ dictionary: [String]) throws -> [String] {
    try DictionaryCore.decodeTrainedDictionaryBlock(block, dictionary)
}

public func applyDictionaryReferences(_ state: SessionState, _ columns: [Column]) {
    DictionaryCore.applyDictionaryReferences(state, columns)
}

public func dictionaryPayloadHash(_ payload: Data) -> UInt64 {
    DictionaryCore.dictionaryPayloadHash(payload)
}
